### PR TITLE
windows wip

### PR DESCRIPTION
**Pull request by nico**
not ready for merging, doesn't even build iirc. also leaks handles, see TODO.

I wrote this 1.5 weeks ago and haven't had time to work more on it since then, so uploading in case someone else wants to continue.

makes subprocesses's stdout and stderr write to the same pipe. that means possibly interleaved output. that matches ninja's behavior, but is worse than n2's non-win behavior which gets stdout and stderr independently from subprocesses and then appends them (so they're guaranteed not to interleave).

(else i'll get back to it sometime soon, hopefully.)

### CODE REVIEW

**Comment by piscisaureus** [comment]
Out of curiosity, what is the reason for reimplementing all this? Rust Command APIs are generally pretty capable.

**Comment by evmar** [comment]
One subtlety I was thinking about in this area is how the `console` pool gives subcommands direct access to the console. 
 Surely the default Rust process abstraction allows that anyway, but in our case we want to show a progress bar while also showing console output or something like that, which might require something trickier.

I think if we could make it work, I'd prefer n2 to not special-case "console" output, in favor of doing more smart-printing logic, like the rendering done here where each spammy command gets its own part of the output stream, maybe?
https://arcan-fe.com/2022/04/02/the-day-of-a-new-command-line-interface-shell/

However, there might be use cases of 'console' where the subcommand wants direct access to the outer stdin, e.g. if it runs `git clone` and that wants to ask for a password or something.

tldr I think using the Rust process abstraction is probably fine.

**Comment by evmar** [comment]
Side note re process output interleaving.

I was previously in Ninja careful to keep stdout and stderr separate, allowing them to interleave, because I was worried about broken programs that write to both and expect the output in roughly the order they were written.  (Even at my dayjob I have been fighting with some programs like this, yuck.)  But programs that mix log writes to stdout and stderr are fundamentally broken because there are potential races when we read from them.   Also it doesn't even make sense to do that.

So I think that overall, just stuffing both stdout/stderr into the same pipe is probably fine, especially because it makes the code simpler.  Even if a program is interleaving writes to both, I think the individual bytes within the `write()`s won't interleave.

**Comment by evmar** [comment]
> Out of curiosity, what is the reason for reimplementing all this? Rust Command APIs are generally pretty capable.

Here's some things I learned from one year in the future!  Rust's command APIs are quite nice, but we end up wanting more subtle things from them.

On Unix: we had the problem where it leaks fds (where Nico slapped a mutex in there, I think maybe you commented on the PR when he did it).  And there's also the behavior I discussed here where we redirect both stdout and stderr into the same pipe, which is relatively easy to do with dup2 but which isn't supported by the command API.  In Ninja we ended up using some more subtle stuff like process groups as well.  And finally my recollection is that switching to use posix_spawn in particular (which is the the higher-level replacement for fork+exec) lets us pass off a bunch of unix details around close-on-exec to an API that handles them.  So recently I switched to using posix_spawn in n2 as well.  It wasn't so bad, I could impl Drop to manage some of the dont-forget-to-cleanup-ness of it.

On Windows, I think(?) we can do the same pipe redirection trick using CreateProcess:

```
        startup_info.hStdOutput = GetStdHandle(STD_OUTPUT_HANDLE);
        startup_info.hStdError = startup_info.hStdOutput;
```

but I haven't verified it.

I write this comment mostly to myself, to reflect on how the main things ninja/n2 do are stat files and spawn processes, so it makes some sense we need to use all the tricks we have for them.

PS: At least for the stat case it seems like `std::fs::metadata(path)` does everything we want, though now that I write this remark I recall something about how at least in C the implementation of `stat()` on Windows has to make multiple system calls to gather all the info expected by that Unixy API.  I guess if I ever really tried to get this going well on Windows I might end up revisiting that part as well.

**Comment by evmar** [comment]
Merge this (with the fancy threadsafe handle inheritance idea) as https://github.com/evmar/n2/commit/0b10c92610495ebb4133f770bcda47b8abc6c89e
